deals/module/deals.go

```diff
@@ -235,61 +235,125 @@ func (m *Module) Watch(ctx context.Context, proposals []cid.Cid) (<-chan deals.D
 	return ch, nil
 }
 
-// FinalDealRecords returns a list of all finalized storage deals.
-// Records are sorted ascending by activation epoch then timestamp.
-func (m *Module) FinalDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getFinalDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting final deals: %v", err)
+// ListStorageDealRecords lists storage deals according to the provided options.
+func (m *Module) ListStorageDealRecords(opts ...deals.ListDealRecordsOption) ([]deals.StorageDealRecord, error) {
+	c := deals.ListDealRecordsConfig{}
+	for _, opt := range opts {
+		opt(&c)
+	}
+	var final []deals.StorageDealRecord
+	if !c.OnlyPending {
+		recs, err := m.store.getFinalDeals()
+		if err != nil {
+			return nil, fmt.Errorf("getting final deals: %v", err)
+		}
+		final = recs
 	}
-	return ret, nil
-}
 
-// PendingDealRecords returns a list of all pending storage deals.
-// Records are sorted ascending by timestamp.
-func (m *Module) PendingDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getPendingDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting pending deals: %v", err)
+	var pending []deals.StorageDealRecord
+	if c.IncludePending || c.OnlyPending {
+		recs, err := m.store.getPendingDeals()
+		if err != nil {
+			return nil, fmt.Errorf("getting pending deals: %v", err)
+		}
+		pending = recs
 	}
-	return ret, nil
-}
 
-// AllDealRecords returns a list of all finalized and pending deals.
-// Records are sorted ascending by activation epoch, if available, then timestamp.
-func (m *Module) AllDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getFinalDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting final deals: %v", err)
-	}
-	pending, err := m.store.getPendingDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting pending deals: %v", err)
+	combined := append(final, pending...)
+
+	var filtered []deals.StorageDealRecord
+
+	if len(c.FromAddrs) > 0 || len(c.DataCids) > 0 {
+		var fromAddrsFilter map[string]struct{}
+		var dataCidsFilter map[string]struct{}
+		for _, addr := range c.FromAddrs {
+			fromAddrsFilter[addr] = struct{}{}
+		}
+		for _, cid := range c.DataCids {
+			dataCidsFilter[cid] = struct{}{}
+		}
+		for _, record := range combined {
+			_, inFromAddrsFilter := fromAddrsFilter[record.Addr]
+			_, inDataCidsFilter := dataCidsFilter[record.DealInfo.PieceCID.String()]
+			includeViaFromAddrs := len(c.FromAddrs) == 0 || inFromAddrsFilter
+			includeViaDataCids := len(dataCidsFilter) == 0 || inDataCidsFilter
```

Comment:
Then decide if each filter contributes to the record being included or not. The answer is yes if the user didn't provide any filter values OR if they did and there's a match.